File: javascript/packages/linter/src/rules/html-no-aria-hidden-on-body.ts
import { ParserRule } from "../types.js"
import { BaseRuleVisitor } from "./rule-utils.js"
import { hasAttribute, getAttributeValue, findAttributeByName, getAttributes, getTagLocalName } from "@herb-tools/core"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { HTMLOpenTagNode, ParseResult } from "@herb-tools/core"

class NoAriaHiddenBodyVisitor extends BaseRuleVisitor {
  visitHTMLOpenTagNode(node: HTMLOpenTagNode): void {
    this.checkAriaHiddenOnBody(node)
    super.visitHTMLOpenTagNode(node)
  }

  private checkAriaHiddenOnBody(node: HTMLOpenTagNode): void {
    const tagName = getTagLocalName(node)

    if (tagName !== "body") return

    if (this.hasAriaHidden(node)) {
      this.addOffense(
        "The `aria-hidden` attribute should never be present on the `<body>` element, as it hides the entire document from assistive technology users.",
        node.tag_name!.location,
      )
    }
  }

  private hasAriaHidden(node: HTMLOpenTagNode): boolean {
    if (!hasAttribute(node, "aria-hidden")) return false

    const attributes = getAttributes(node)
    const ariaHiddenAttr = findAttributeByName(attributes, "aria-hidden")

    if (!ariaHiddenAttr) return false

    const value = getAttributeValue(ariaHiddenAttr)

    return value === null || value === "" || value === "true"
  }
}

export class HTMLNoAriaHiddenOnBodyRule extends ParserRule {
  static ruleName = "html-no-aria-hidden-on-body"
  static introducedIn = this.version("0.9.0")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "warning"
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new NoAriaHiddenBodyVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/html_no_aria_hidden_on_body.rs
use crate::utils::tag_utils::{get_tag_name_from_open_tag, has_aria_hidden, tag_name_location};

use herb::nodes::HTMLOpenTagNode;
use herb::Visitor;

rule_visitor!(NoAriaHiddenOnBodyVisitor);
define_parser_rule!(HTMLNoAriaHiddenOnBodyRule, "html-no-aria-hidden-on-body", Error, NoAriaHiddenOnBodyVisitor);

impl Visitor for NoAriaHiddenOnBodyVisitor {
  fn visit_html_open_tag_node(&mut self, node: &HTMLOpenTagNode) {
    if let Some(tag_name) = get_tag_name_from_open_tag(node) {
      if tag_name.eq_ignore_ascii_case("body") && has_aria_hidden(node) {
        self.add_offense(
          "The `aria-hidden` attribute should never be present on the `<body>` element, as it hides the entire document from assistive technology users."
            .to_string(),
          tag_name_location(node),
        );
      }
    }

    self.walk_html_open_tag_node(node);
  }
}
